core/src/cluster_info_repair_listener.rs

```diff
@@ -0,0 +1,683 @@
+                    }
+                }
+            }
+
+            // After updating all the peers, send out repairs to those that need it
+            let _ = Self::serve_repairs(
+                &my_id,
+                blocktree,
+                peer_roots,
+                &peers_needing_repairs,
+                &socket,
+                cluster_info,
+                &epoch_schedule,
+                &mut my_gossiped_root,
+            );
+
+            sleep(Duration::from_millis(REPAIRMEN_SLEEP_MILLIS as u64));
+        }
+    }
+
+    fn serve_repairs(
+        my_id: &Pubkey,
+        blocktree: &Blocktree,
+        peer_roots: &HashMap<Pubkey, (u64, u64)>,
+        repairees: &HashMap<Pubkey, EpochSlots>,
+        socket: &UdpSocket,
+        cluster_info: &Arc<RwLock<ClusterInfo>>,
+        epoch_schedule: &EpochSchedule,
+        my_gossiped_root: &mut u64,
+    ) -> Result<()> {
+        for (repairee_id, repairee_epoch_slots) in repairees {
+            let repairee_root = repairee_epoch_slots.root;
+
+            let repairee_tvu = {
+                let r_cluster_info = cluster_info.read().unwrap();
+                let contact_info = r_cluster_info.get_contact_info_for_node(repairee_id);
+                contact_info.map(|c| c.tvu)
+            };
+
+            if let Some(repairee_tvu) = repairee_tvu {
+                // For every repairee, get the set of repairmen who are responsible for
+                let mut eligible_repairmen = Self::find_eligible_repairmen(
+                    repairee_root,
+                    peer_roots,
+                    epoch_schedule,
+                    NUM_BUFFER_SLOTS,
+                );
+
+                eligible_repairmen.push(my_id);
+
+                Self::shuffle_repairmen(
+                    &mut eligible_repairmen,
+                    repairee_id,
+                    repairee_epoch_slots.root,
+                );
+
+                let my_root = Self::update_my_gossiped_root(my_id, cluster_info, my_gossiped_root);
+
+                let _ = Self::serve_repairs_to_repairee(
+                    my_id,
+                    my_root,
+                    blocktree,
+                    &repairee_epoch_slots,
+                    &eligible_repairmen,
+                    socket,
+                    &repairee_tvu,
+                );
+            }
+        }
+
+        Ok(())
+    }
+
+    fn serve_repairs_to_repairee(
+        my_id: &Pubkey,
+        my_root: u64,
+        blocktree: &Blocktree,
+        repairee_epoch_slots: &EpochSlots,
+        eligible_repairmen: &[&Pubkey],
+        socket: &UdpSocket,
+        repairee_tvu: &SocketAddr,
+    ) -> Result<()> {
+        let slot_iter = blocktree
+            .slot_meta_iterator(repairee_epoch_slots.root + 1)
+            .expect("Couldn't get db iterator");
+        let mut total_data_blobs_sent = 0;
+        let mut total_coding_blobs_sent = 0;
+
+        for (slot, slot_meta) in slot_iter {
```

Comment:
repairmen are responsible for exactly one slot of blobs per gossip update for the peer

Reply:
ah ok, makes sense, so repairmen first have to divide up responsibility for the missing slots, and then also the blobs within those slots?

Reply:
there should be exactly one next slot inferrable from the repairee's request...